Modules/_collectionsmodule.c

```diff
@@ -2328,6 +2328,107 @@ _count_elements(PyObject *self, PyObject *args)
     Py_RETURN_NONE;
 }
 
+/* Helper functions for namedtuples */
+
+typedef struct {
+    propertyobject po;
+    Py_ssize_t index;
+    PyObject* doc;
+} _tuplegetterobject;
+
+static PyObject *
+tuplegetter_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    Py_ssize_t index;
+    PyObject *doc;
+    _tuplegetterobject *self;
+
+    static char *kwlist[] = {"index", "doc", NULL};
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "n$O", kwlist,
+                                     &index, &doc)){
+        return NULL;
+    }
+
+    self = (_tuplegetterobject *)PyProperty_Type.tp_new(type, args, kwds);
+    if (self == NULL) {
+        return NULL;
+    }
+    self->index = index;
+    Py_INCREF(doc);
+    self->doc = doc;
+    return (PyObject *)self;
+}
+
+static PyObject *
+tuplegetterdescr_get(PyObject *self, PyObject *obj, PyObject *type)
+{
+    PyObject *result;
+    if (obj == NULL || obj == Py_None) {
+        Py_INCREF(self);
+        return self;
+    }
+    result = PyTuple_GetItem(obj, ((_tuplegetterobject*)self)->index);
```

Comment:
We could add a Pytuple_CheckExact here, but as this class is private to the collections module,
we can assume that the contract is that it has to be used in a tuple-like class. Notice that this stills throws `SystemError: Objects/tupleobject.c:152: bad argument to internal function` if is used incorrectly.

Reply:
If this can be triggered from Python, we should avoid `SystemError` and crashes.

Reply:
Made a `PyTuple_Check` that throws a `TypeError` in 21be735.